pkg/component/controller/etcd_member_reconciler.go

```diff
@@ -0,0 +1,316 @@
+	"github.com/k0sproject/k0s/pkg/component/controller/leaderelector"
+	"github.com/k0sproject/k0s/pkg/component/manager"
+	"github.com/k0sproject/k0s/pkg/config"
+	"github.com/k0sproject/k0s/pkg/etcd"
+	kubeutil "github.com/k0sproject/k0s/pkg/kubernetes"
+	"github.com/k0sproject/k0s/pkg/kubernetes/watch"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	extensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	extclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+)
+
+var _ manager.Component = (*EtcdMemberReconciler)(nil)
+
+// TODO: DO we need REady probing for this component?
+// var _ manager.Ready = (*EtcdMemberReconciler)(nil)
+
+func NewEtcdMemberReconciler(kubeClientFactory kubeutil.ClientFactoryInterface, k0sVars *config.CfgVars, etcdConfig *v1beta1.EtcdConfig, leaderElector leaderelector.Interface) (*EtcdMemberReconciler, error) {
+
+	return &EtcdMemberReconciler{
+		clientFactory: kubeClientFactory,
+		k0sVars:       k0sVars,
+		etcdConfig:    etcdConfig,
+		leaderElector: leaderElector,
+	}, nil
+}
+
+type EtcdMemberReconciler struct {
+	clientFactory    kubeutil.ClientFactoryInterface
+	k0sVars          *config.CfgVars
+	etcdConfig       *v1beta1.EtcdConfig
+	etcdMemberClient etcdmemberclient.EtcdMemberInterface
+	leaderElector    leaderelector.Interface
+}
+
+func (e *EtcdMemberReconciler) Init(_ context.Context) error {
+	return nil
+}
+
+func (e *EtcdMemberReconciler) Start(ctx context.Context) error {
+	log := logrus.WithField("component", "etcdMemberReconciler")
+
+	etcdMemberClient, err := e.clientFactory.GetEtcdMemberClient()
+	if err != nil {
+		return err
+	}
+	e.etcdMemberClient = etcdMemberClient
+
+	// Run the watch in go routine so it keeps running till the context ends
+	go func() {
+		err = e.waitForCRD(ctx)
+		if err != nil {
+			log.WithError(err).Errorf("didn't see EtcdMember CRD ready in time")
+			return
+		}
+
+		// Create the object for this node
+		err = e.createMemberObject()
+		if err != nil {
+			log.WithError(err).Error("failed to create EtcdMember object")
+		}
+		var lastObservedVersion string
+		err = watch.EtcdMembers(etcdMemberClient).
+			WithErrorCallback(func(err error) (time.Duration, error) {
+				retryDelay, e := watch.IsRetryable(err)
+				if e == nil {
+					log.WithError(err).Debugf(
+						"Encountered transient error while watching etcd members"+
+							", last observed resource version was %q"+
+							", retrying in %s",
+						lastObservedVersion, retryDelay,
+					)
+					return retryDelay, nil
+				}
+				log.WithError(e).Error("bailing out watch")
+				return 0, err
+			}).
+			IncludingDeletions().
+			Until(ctx, func(member *etcdv1beta1.EtcdMember) (bool, error) {
+				e.reconcileMember(ctx, member)
+				// Never stop the watch
+				return false, nil
+			})
+		if err != nil {
+			log.WithError(err).Warn("watch terminated")
```

Comment:
I reckon this will log an error, even on coordinated shutdown.

Reply:
I know, hence warn level only. Dunno if it makes sense to try to differentiate between a gracefull shutdown and abnormal termination